base_layer/wallet/src/util/encryption.rs

```diff
@@ -0,0 +1,70 @@
+// Copyright 2020. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use aes_gcm::{
+    aead::{generic_array::GenericArray, Aead, Error as AeadError},
+    Aes256Gcm,
+};
+use rand::{rngs::OsRng, RngCore};
+
+pub const AES_NONCE_BYTES: usize = 12;
+pub const AES_KEY_BYTES: usize = 32;
+
+pub trait Encryptable<C> {
+    fn encrypt(&mut self, cipher: &C) -> Result<(), AeadError>;
+    fn decrypt(&mut self, cipher: &C) -> Result<(), AeadError>;
+}
+
+pub fn decrypt_bytes_integral_nonce(cipher: &Aes256Gcm, mut ciphertext: Vec<u8>) -> Result<Vec<u8>, AeadError> {
+    let data = ciphertext.split_off(AES_NONCE_BYTES);
+    let nonce = GenericArray::from_slice(ciphertext.as_slice());
+    cipher.decrypt(nonce, data.as_ref())
```

Comment:
nit: a little cleaner?
```suggestion
  let (nonce_bytes, cipher_text) = ciphertext.split_at(AES_NONCE_BYTES);
    let nonce = GenericArray::from_slice(nonce_bytes);
    cipher.decrypt(nonce, cipher_text)
```

Reply:
Bit concerned that the input is not checked anywhere that I can see. `split_at` (and `split_off`) will panic if the input is no long enough, so you'd need to check either in the calling code or here for the correct data length

Reply:
Def, I remember putting it in there but must of fallen out along the way.